pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureEnvironmentProvider.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.commons.lang.StringUtils;
+import org.apache.http.HttpEntityEnclosingRequest;
+import org.apache.http.HttpStatus;
+import org.apache.http.StatusLine;
+import org.apache.http.client.HttpRequestRetryHandler;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.environmentprovider.PinotEnvironmentProvider;
+
+
+/**
+ * Azure Environment Provider used to retrieve azure cloud specific instance configuration.
+ */
+public class AzureEnvironmentProvider implements PinotEnvironmentProvider {
+
+  protected static final String MAX_RETRY = "maxRetry";
+  protected static final String IMDS_ENDPOINT = "imdsEndpoint";
+  protected static final String CONNECTION_TIMEOUT_MILLIS = "connectionTimeoutMillis";
+  protected static final String REQUEST_TIMEOUT_MILLIS = "requestTimeoutMillis";
+  private static final String COMPUTE = "compute";
+  private static final String METADATA = "Metadata";
+  private static final String PLATFORM_FAULT_DOMAIN = "platformFaultDomain";
+  private int _maxRetry;
+  private String _imdsEndpoint;
+  private CloseableHttpClient _closeableHttpClient;
+
+  public AzureEnvironmentProvider() {
+  }
+
+  public void init(PinotConfiguration pinotConfiguration) {
+    Preconditions.checkArgument(0 < Integer.parseInt(pinotConfiguration.getProperty(MAX_RETRY)),
+        "[AzureEnvironmentProvider]: " + MAX_RETRY + " cannot be less than or equal to 0");
+    Preconditions.checkArgument(!StringUtils.isBlank(pinotConfiguration.getProperty(IMDS_ENDPOINT)),
+        "[AzureEnvironmentProvider]: " + IMDS_ENDPOINT + " should not be null or empty");
+
+    _maxRetry = Integer.parseInt(pinotConfiguration.getProperty(MAX_RETRY));
+    _imdsEndpoint = pinotConfiguration.getProperty(IMDS_ENDPOINT);
+    int connectionTimeoutMillis = Integer.parseInt(pinotConfiguration.getProperty(CONNECTION_TIMEOUT_MILLIS));
+    int requestTimeoutMillis = Integer.parseInt(pinotConfiguration.getProperty(REQUEST_TIMEOUT_MILLIS));
+
+    final RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(connectionTimeoutMillis)
+        .setConnectionRequestTimeout(requestTimeoutMillis).build();
+
+    final HttpRequestRetryHandler httpRequestRetryHandler =
+        (iOException, executionCount, httpContext) -> !(executionCount >= _maxRetry
+            || iOException instanceof InterruptedIOException || iOException instanceof UnknownHostException
+            || iOException instanceof SSLException || HttpClientContext.adapt(httpContext)
+            .getRequest() instanceof HttpEntityEnclosingRequest);
+
+    _closeableHttpClient =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *\_closeableHttpClient*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).